### DISCUSSION TITLE
Add a generalized `%unique`

### DISCUSSION THREAD

**Discussion by eternaleye** [beetbox/beets#5110] [Ideas]
One thing I've found myself wanting is to be able to conditionally add disambiguators in paths at granularities other than the album or singleton level. I think a generalized `%unique` path format function would be helpful for this; it could work like so:

Given `%unique{<keys>,<fields>,<disambiguators>,<brackets>}`, it will resolve to nothing if the values of `keys` are sufficient to resolve the tuple of `fields` to a unique value. If not, it will add fields from `disambiguators`, enclosed in `brackets`, until that condition holds.

This is strictly more expressive than the existing `%aunique` and `%sunique` functions, as (IIUC) they can be implemented atop it:

- `%aunique{<keys>,<disambiguators>,<brackets>}` => `%unique{<keys>,album_id,<disambiguators> album_id,<brackets>}`
- `%sunique{<keys>,<disambiguators>,<brakcets>}` => `%unique{<keys> album_id singleton,id,<disambiguators> id,<brackets>}`

In my particular case, I want to separate distinct artists with a common name, so I want `%unique{albumartist,mb_albumartistid,mb_albumartistid}` in my album paths and `%unique{artist,mb_artistid,mb_artistid}` in my singleton paths.

I also nest releases under folders for the release group, so I further want my album paths to have:

- `${original_fulldate} -- $album %unique{albumartist original_fulldate release_group_title,mb_releasegroupid,mb_releasegroupid}`

(Given this, it may also be good for `disambiguators` to default to `fields` if not present, simplifying the above to `%unique{albumartist,mb_albumartistid}` and `%unique{artist,mb_artistid}`)

**Comment by michaelkrieger** [comment]
**@eternaleye : Within 'latest' there is Singleton Disambiguation:**

> It is also possible to have singleton tracks with the same name and the same artist. Beets provides the %sunique{} template to avoid giving these tracks the same file path. It has the same arguments as the [%aunique](https://beets.readthedocs.io/en/latest/reference/pathformat.html#aunique) template, but the default values are different. The default identifiers are artist title and the default disambiguators are year trackdisambig.

**I'd love, however for it to be able to use an autoincrement to make unique paths:**
Lets say I have two copies of the single song "Artist - Song".  They're different versions of the song and frankly just need to be named better (ie: Club Remix or Album Version or Radio Edit) but I don't yet know the detailed track information.  These are often historical files and used in playlists.  So bringing them all in, I'd want a %unique as follows:
`singleton: Singles/$year/$artist - $title%unique(year, artist, title, '()')`
My expectation is that I could get:
`Singles/2024/Artist - Title` and `Singles/2024/Artist - Title (2)` and `Singles/2024/Artist - Title (3)` out of three tracks.
At this point I have to set trackdisambig for each track.

**Reply by eternaleye** [reply]
I was under the impression you could use `id`, the beets library database numeric primary key, for this? Similar to how albums with `%aunique` fall back to the `album_id` number. It's not quite an autoincrement, but the semantics of an autoincrement are... somewhat fraught, because if your path template changes you'd need to assign a canonical _ordering_ to the _new_ set of tracks it conflicts with.
